{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First Principles Approach To Solving My Graduation Project\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Current Suggested Approach\n",
    "\n",
    "1. **What the model sees as input:**  \n",
    "   The current design of the model takes two things:\n",
    "   - A representation of how the cell’s gene expression changes when exposed to a particular drug. In other words, you give the model a snapshot of the cell’s transcriptional profile after it has already been treated.\n",
    "   - The chemical descriptors (or structure) of the drug itself.\n",
    "\n",
    "2. **What the model outputs:**  \n",
    "   Given these two pieces of information—the observed biological response of the cell’s transcriptome to the drug and the drug’s own structural characteristics—the model predicts how viable the cell remains. Low viability means the drug killed the cancer cells effectively; high viability means it did not.\n",
    "\n",
    "3. **What this means in practice:**  \n",
    "   This setup effectively lets the model learn the following relationship:  \n",
    "\n",
    "   *\"Whenever an **arbitrary** cell line shows this specific pattern of gene expression change after being treated with a drug that has these particular molecular features, we often see a certain level of cell death or survival.\"*\n",
    "\n",
    "   In other words, the model is learning a mapping:\n",
    "   $$\n",
    "   f(\\text{Drug Structure}, \\text{Post-Treatment Gene Expression Changes}) \\rightarrow \\text{Viability}.\n",
    "   $$\n",
    "\n",
    "   This is a correlation-based, supervised learning problem where the model leverages direct evidence of how the drug affects the cells at the molecular level (transcriptome) and combines it with the drug’s own intrinsic properties.\n",
    "\n",
    "4. **From first principles, what knowledge does the model acquire?**  \n",
    "   - The model learns which gene expression patterns (induced by a given drug) are associated with higher or lower cell survival. It’s essentially learning the \"fingerprints\" of effective vs. ineffective drug-induced perturbations on the cell’s biology.\n",
    "   - It also learns how certain drug structures tend to produce particular transcriptional responses. Over many training examples, it can infer that drugs with certain chemical traits consistently lead to specific transcriptional signatures, which in turn correlate with certain viability outcomes.\n",
    "\n",
    "   In essence, the model is encoding a conditional relationship: *If a drug looks like this (chemically) and produces these changes in gene expression, then the cell viability will be such and such.*  \n",
    "\n",
    "5. **Why this might not achieve the ultimate goal yet:**  \n",
    "   The main practical aim is to predict whether a new drug will kill tumor cells **without first testing it in the lab**. However, the current model needs to be given the gene expression changes caused by that drug on that cell line, which is something you only get after you've already exposed the cells to the drug. This is a catch-22: to predict viability, the model relies on the very data (post-treatment transcriptomics) that you wouldn't have before testing.\n",
    "\n",
    "   So while the model “learns” a sophisticated correlation between (drug structure + observed cellular response) and viability, it doesn't learn to predict viability from only the drug and baseline cell information. It requires the post-perturbation expression data, effectively making it a \"reactive\" predictor rather than a \"proactive\" one. It’s like the model is very good at saying “Given I see these molecular changes after the drug has done its work, here’s the outcome,” rather than “Given this drug and this cell line, what would happen if I apply the drug?”\n",
    "\n",
    "6. **In conclusion:**  \n",
    "   The current model, as described, fundamentally learns a mapping from *observed drug-induced cellular changes* (transcriptomic perturbation) combined with the *drug’s intrinsic properties* to a final viability outcome. It does not inherently learn how to predict viability without the intermediate step of knowing the gene expression changes induced by the drug. In first-principles terms, it is a function that takes a known effect (the transcriptomic change) and a cause (the drug) and predicts the final outcome (viability), rather than predicting the outcome directly from just the cause and the baseline conditions alone.\n",
    "\n",
    "This means the model is not yet solving the intended problem of predicting efficacy before any testing is done. It's learning a useful relationship, but one that still depends on having post-treatment data that you wouldn’t have in a real predictive scenario.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building A New Approach\n",
    "\n",
    "1. **Thinking in Terms of Vector Spaces:**\n",
    "   You start with a baseline gene expression vector that represents the cell line’s \"state.\" A drug can be thought of as an operator or transformation that maps this baseline vector into a new vector representing the cell's perturbed state. If the drug is effective, it ideally pushes (or \"rotates\") this state vector into a region of the state space associated with low viability (i.e., the cell is likely to die).\n",
    "\n",
    "   In this analogy:\n",
    "   - **Baseline State Vector:** Encodes the cell line’s intrinsic biological configuration under normal conditions.\n",
    "   - **Drug:** Acts like a transformation (linear or nonlinear) that takes the baseline vector to a new \"perturbed\" vector.\n",
    "   - **Viability:** A function that assigns a \"score\" or \"label\" to each point in the gene expression space. Some regions correspond to healthy, viable cells; other regions correspond to stressed or dying cells.\n",
    "\n",
    "![Vector Intuition](../assets/images/Intuition.png)\n",
    "\n",
    "2. **Is the Transformation Universal?**\n",
    "   One might imagine a perfect world where the same drug always moves the vector from a \"viable region\" to a \"non-viable region.\" But in reality:\n",
    "   - The effect of the drug (the \"rotation\") depends on both the drug’s properties and **the cell line’s baseline biology**.\n",
    "   - Different cell lines, even starting from similar baseline states, may not respond identically to the same drug. The transformation induced by the drug can be cell-line specific because the drug interacts with the particular molecular networks, receptors, and pathways unique to that cell line.\n",
    "\n",
    "   Thus, the \"map\" from baseline → perturbed state → viability is not one-size-fits-all. The drug’s effect is **context-dependent** and can vary significantly across different cell lines.\n",
    "\n",
    "3. **Different Viability Reactions to the Same Final State:**\n",
    "   Consider the question: If two different cell lines, through different means, arrive at the same final gene expression vector, do they have the same viability?\n",
    "\n",
    "   In theory, if the exact final state vector is identical and fully captures all relevant biology, and the viability function is solely a function of that final state vector, then yes, they would have the same viability. But in practice:\n",
    "   - The viability outcome might also depend on contextual, cell-line-specific factors not fully captured by the gene expression vector alone. For example, gene expression is not the only determinant of cell fate. Epigenetic context, protein modifications, metabolic states, and other hidden variables could differ between cell lines.\n",
    "   - The model may implicitly encode cell-line identity in the baseline features. If baseline differences are part of the input, then the same \"final state vector\" might not actually be interpreted identically by the model for two different cell lines.\n",
    "\n",
    "   Essentially, the final viability is not just a function of the final gene expression vector alone, but also of the cell line context represented in the baseline state. Two cell lines may map the same final gene expression pattern to different viability outcomes if they differ in some underlying, unmodeled dimensions or if the model includes baseline identifiers that change how it interprets that final state.\n",
    "\n",
    "4. **Conclusion:**\n",
    "   - Conceptually, you can think of the drug as something that tries to move the baseline gene expression vector into a \"low-viability\" region.\n",
    "   - However, each cell line has its own \"viability landscape.\" The same final gene expression pattern might not yield the same viability across different cell lines if other contextual factors differ.\n",
    "   - In practice, the model (and biological reality) is more complicated than a simple uniform map of state vectors to viability. The cell line context, baseline conditions, and unobserved variables mean that different cell lines can indeed have different viability reactions to a given \"state vector.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\20191678\\AppData\\Local\\Temp/ipykernel_27364/2645538182.py:3: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"../data/raw/instinfo_beta.txt\",sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original shape: (3026460, 30)\n",
      "Shape after filtering: (1312170, 30)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"../data/raw/instinfo_beta.txt\",sep=\"\\t\")\n",
    "\n",
    "# Only select rows where the column qc_pass is 1.0 and where the pert_type is trt_cp\n",
    "df_qc = df[(df[\"qc_pass\"] == 1.0) & (df[\"pert_type\"] == \"trt_cp\")]\n",
    "\n",
    "# Print shapes to compare\n",
    "print(\"Original shape:\", df.shape)\n",
    "print(\"Shape after filtering:\", df_qc.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bead_batch</th>\n",
       "      <th>nearest_dose</th>\n",
       "      <th>pert_dose</th>\n",
       "      <th>pert_dose_unit</th>\n",
       "      <th>pert_idose</th>\n",
       "      <th>pert_time</th>\n",
       "      <th>pert_itime</th>\n",
       "      <th>pert_time_unit</th>\n",
       "      <th>cell_mfc_name</th>\n",
       "      <th>pert_mfc_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>pert_type</th>\n",
       "      <th>cell_iname</th>\n",
       "      <th>qc_pass</th>\n",
       "      <th>dyn_range</th>\n",
       "      <th>inv_level_10</th>\n",
       "      <th>build_name</th>\n",
       "      <th>failure_mode</th>\n",
       "      <th>project_code</th>\n",
       "      <th>cmap_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.00</td>\n",
       "      <td>uL</td>\n",
       "      <td>20 uL</td>\n",
       "      <td>72.0</td>\n",
       "      <td>72 h</td>\n",
       "      <td>h</td>\n",
       "      <td>VCAP</td>\n",
       "      <td>ERG_11</td>\n",
       "      <td>...</td>\n",
       "      <td>ERG013_VCAP_72H_X3_B11:O14</td>\n",
       "      <td>trt_sh</td>\n",
       "      <td>VCAP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.20788</td>\n",
       "      <td>4220.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dyn_range</td>\n",
       "      <td>ERG</td>\n",
       "      <td>ERG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>uL</td>\n",
       "      <td>1 uL</td>\n",
       "      <td>96.0</td>\n",
       "      <td>96 h</td>\n",
       "      <td>h</td>\n",
       "      <td>U2OS</td>\n",
       "      <td>TRCN0000072237</td>\n",
       "      <td>...</td>\n",
       "      <td>TAK004_U2OS_96H_X2_B10_DUO52HI53LO:D10</td>\n",
       "      <td>ctl_vector</td>\n",
       "      <td>U2OS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.73906</td>\n",
       "      <td>1462.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inv_level_10</td>\n",
       "      <td>TAK</td>\n",
       "      <td>LACZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.10</td>\n",
       "      <td>ng/ml</td>\n",
       "      <td>0.1 ng/ml</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2 h</td>\n",
       "      <td>h</td>\n",
       "      <td>HEPG2</td>\n",
       "      <td>SOD3</td>\n",
       "      <td>...</td>\n",
       "      <td>CYT001_HEPG2_2H_X2_B12:N12</td>\n",
       "      <td>trt_lig</td>\n",
       "      <td>HEPG2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.79642</td>\n",
       "      <td>3038.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYT</td>\n",
       "      <td>SOD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150.00</td>\n",
       "      <td>ng</td>\n",
       "      <td>150 ng</td>\n",
       "      <td>48.0</td>\n",
       "      <td>48 h</td>\n",
       "      <td>h</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>ENTRY00543</td>\n",
       "      <td>...</td>\n",
       "      <td>HSF038_HEK293T_48H_X2_B12:M01</td>\n",
       "      <td>trt_oe</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.79710</td>\n",
       "      <td>1642.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inv_level_10</td>\n",
       "      <td>HSF</td>\n",
       "      <td>PDGFRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>f3b5</td>\n",
       "      <td>6.66</td>\n",
       "      <td>5.33</td>\n",
       "      <td>uM</td>\n",
       "      <td>6.66 uM</td>\n",
       "      <td>24.0</td>\n",
       "      <td>24 h</td>\n",
       "      <td>h</td>\n",
       "      <td>A375</td>\n",
       "      <td>BRD-K79781870</td>\n",
       "      <td>...</td>\n",
       "      <td>DOS043_A375_24H_X1_F3B5_DUO52HI53LO:D17</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>A375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.78867</td>\n",
       "      <td>1558.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inv_level_10,qc_iqr</td>\n",
       "      <td>DOS</td>\n",
       "      <td>BRD-K79781870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  bead_batch  nearest_dose  pert_dose pert_dose_unit pert_idose  pert_time  \\\n",
       "0        b11           NaN      20.00             uL      20 uL       72.0   \n",
       "1        b10           NaN       1.00             uL       1 uL       96.0   \n",
       "2        b12           NaN       0.10          ng/ml  0.1 ng/ml        2.0   \n",
       "3        b12           NaN     150.00             ng     150 ng       48.0   \n",
       "4       f3b5          6.66       5.33             uM    6.66 uM       24.0   \n",
       "\n",
       "  pert_itime pert_time_unit cell_mfc_name     pert_mfc_id  ...  \\\n",
       "0       72 h              h          VCAP          ERG_11  ...   \n",
       "1       96 h              h          U2OS  TRCN0000072237  ...   \n",
       "2        2 h              h         HEPG2            SOD3  ...   \n",
       "3       48 h              h       HEK293T      ENTRY00543  ...   \n",
       "4       24 h              h          A375   BRD-K79781870  ...   \n",
       "\n",
       "                                 sample_id   pert_type cell_iname qc_pass  \\\n",
       "0               ERG013_VCAP_72H_X3_B11:O14      trt_sh       VCAP     0.0   \n",
       "1   TAK004_U2OS_96H_X2_B10_DUO52HI53LO:D10  ctl_vector       U2OS     0.0   \n",
       "2               CYT001_HEPG2_2H_X2_B12:N12     trt_lig      HEPG2     1.0   \n",
       "3            HSF038_HEK293T_48H_X2_B12:M01      trt_oe    HEK293T     0.0   \n",
       "4  DOS043_A375_24H_X1_F3B5_DUO52HI53LO:D17      trt_cp       A375     0.0   \n",
       "\n",
       "   dyn_range  inv_level_10  build_name         failure_mode  project_code  \\\n",
       "0    4.20788        4220.5         NaN            dyn_range           ERG   \n",
       "1    4.73906        1462.0         NaN         inv_level_10           TAK   \n",
       "2    6.79642        3038.0         NaN                  NaN           CYT   \n",
       "3   23.79710        1642.0         NaN         inv_level_10           HSF   \n",
       "4    6.78867        1558.0         NaN  inv_level_10,qc_iqr           DOS   \n",
       "\n",
       "       cmap_name  \n",
       "0            ERG  \n",
       "1           LACZ  \n",
       "2           SOD3  \n",
       "3         PDGFRA  \n",
       "4  BRD-K79781870  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "5ARG45",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
